Lib/exception_group.py

```diff
@@ -0,0 +1,215 @@
+            indent += 4
+            for e in exc.excs:
+                t = tbg.tb_next_map[id(e)]
+                output.append('---------------------------------------')
+                output.extend(ExceptionGroup.render(e, t, indent))
+        for l in output:
+            print(l)
+        return output
+
+    def __iter__(self):
+        ''' iterate over the individual exceptions (flattens the tree) '''
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup):
+                for e_ in e:
+                    yield e_
+            else:
+                yield e
+
+    def __len__(self):
+        l = 0
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup):
+                l += len(e)
+            else:
+                l += 1
+        return l
+
+    def __repr__(self):
+        return f"ExceptionGroup({self.excs})"
+
+    @staticmethod
+    def catch(types, handler):
+        return ExceptionGroupCatcher(types, handler)
+
+class ExceptionGroupCatcher:
+    """ Based on trio.MultiErrorCatcher """
+
+    def __init__(self, types, handler):
+        """ Context manager to catch and handle ExceptionGroups
+
+        types: the exception types that this handler is interested in
+        handler: a function that takes an ExceptionGroup of the
+        matched type and does something with them
+
+        Any rest exceptions are raised at the end as another
+        exception group
+        """
+        self.types = types
+        self.handler = handler
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, etype, exc, tb):
+        if exc is not None and isinstance(exc, ExceptionGroup):
+            match, rest = exc.split(self.types)
+
+            if not match:
+                # Let the interpreter reraise the exception
+                return False
+
+            handler_excs = self.handler(match)
+            if handler_excs == match:
+                # handler reraised all of the matched exceptions.
+                # reraise exc as is.
+                return False
+
+            if not handler_excs and not rest:
+                # handled and swallowed all exceptions
+                # do not raise anything.
+                return True
+
+            if not rest:
+                to_raise = handler_excs  # raise what handler returned
+            elif not handler_excs:
+                to_raise = rest       # raise the rest exceptions
+            else:
+                # to_keep: EG subgroup of exc with only those to reraise
+                # (either not matched or reraised by handler)
+                to_keep = exc.subgroup(
+                    list(rest) + [e for e in handler_excs if e in match])
+                # to_add: new exceptions raised by handler
+                to_add = handler_excs.subgroup(
+                    [e for e in handler_excs if e not in match])
+                if to_add:
+                    to_raise = ExceptionGroup([to_keep, to_add])
```

Comment:
This introduces a spurious extra nesting level. It’s why I was asking for a merge operator.

Reply:
Hm, that's a tough one. Suppose you have:
```
def f():
    try:
        ...something that raises EG([E1, E2]) or perhaps EG([E2])...
    except *E1:
        ...handler code...
```
Then what bubbles out of `f()` is `EG([E2])`:

- if the try block raises `EG([E1, E2])` then what bubbles out is the remainder, `EG([E2])`
- if the try block raises `EG([E2])` then what bubbles out is exactly that

But if the handler raises `E3`, what bubbles out is wrapped in an extra `EG`, and `E3` is wrapped in *another* extra `EG`:
```
EG(EG([E2]), EG([E3]))
```
Note that the latter is apparently the responsibility of the handler -- you blatantly assume that `handler_excs` has a `subgroup()` method, i.e. that it is an `EG`.

I had somehow hoped that we'd get this instead:
```
EG([E2, E3])
```
But I see we can't do that because the tracebacks are different. I guess it's just something users who are manually unpacking the `EG` nesting structure will have to be aware of -- users using `except *` will be oblivious of this detail.

I'm a little sad about this because I had hoped that the `EG` nesting structure would reflect the TaskGroup nesting structure exactly, but clearly this cannot always be preserved. Oh well.

Reply:
Exactly. 

I was going to suggest that we allow the handler to return a list of exceptions, and then the result can be `EG(EG([E2]), E3))`.  We need to allow both exception list and EG, in case the handler actually did catch an EG with a traceback and so on.

Reply:
Perhaps the handler can just raise exceptions instead of returning them?

Reply:
Then the handler appears in the traceback, and I think that might make sense for new exceptions but not for ones it reraises.  Maybe we should allow both, because raising is very natural for new exceptions.

A use case that Nathaniel and Yuri were talking about was filtering OSErrors - the handler takes all OSErrors, checks their error code, handles one of them an tells us to reraise the rest.

Reply:
Hm, but this thinking is geared towards the interim situation where users have to write a handler function. In the bright future we'll do away with handlers and they just use `except *E`. If that raises an EG it will appear in the traceback.

Realistically users will write broken handler functions that sometimes raise anyway, so we'll have to catch exceptions from the handler. A handler that truly re-raises can do so using bare `raise` -- IMO for everything else it's fine to have the `raise` statement appear in the traceback.

Reply:
So the handler has some return value that means "naked raise", and it raises in all other cases?